mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreRightManager.java

```diff
@@ -161,12 +163,45 @@ public void setRights(MailboxId mailboxId, MailboxACL mailboxACL, MailboxSession
 
     @Override
     public void setRights(MailboxPath mailboxPath, MailboxACL mailboxACL, MailboxSession session) throws MailboxException {
+        validateDomainsAreIdentical(mailboxPath.getUser(), mailboxACL);
         MailboxMapper mapper = mailboxSessionMapperFactory.getMailboxMapper(session);
         Mailbox mailbox = mapper.findMailboxByPath(mailboxPath);
 
         setRights(mailboxACL, mapper, mailbox);
     }
 
+    @VisibleForTesting
+    void validateDomainsAreIdentical(String user, MailboxACL mailboxACL) throws DifferentDomainException {
+        if (mailboxACL.getEntries().keySet().stream()
+            .filter(entry -> !entry.getNameType().equals(NameType.special))
+            .map(EntryKey::getName)
+            .anyMatch(name -> areDomainsDifferent(name, user))) {
+            throw new DifferentDomainException();
+        }
+    }
+
+    @VisibleForTesting
+    boolean areDomainsDifferent(String user, String otherUser) {
+        Optional<String> domain = extractDomain(user);
+        Optional<String> otherDomain = extractDomain(otherUser);
+        if (domain.isPresent() != otherDomain.isPresent()) {
```

Comment:
Optional implements equality the right way : 
```
    public boolean equals(Object var1) {
        if (this == var1) {
            return true;
        } else if (!(var1 instanceof Optional)) {
            return false;
        } else {
            Optional var2 = (Optional)var1;
            return Objects.equals(this.value, var2.value);
        }
    }
```
So a simple `return extractDomain(user).equals(extractDomain(otherUser))` should work